src/control/lib/netdetect/netdetect.go

```diff
@@ -212,6 +211,50 @@ func (da *DeviceAffinity) String() string {
 	return fmt.Sprintf("%s:%s:%s:%d", da.DeviceName, da.CPUSet, da.NodeSet, da.NUMANode)
 }
 
+type hwlocProtectedAccess struct {
```

Comment:
A good start, but I'm not sure I'd limit this interface to only things we need to mutex access to. I was thinking more like hwloc Go bindings for the control plane to use, which would encapsulate all the locking, any complex bookkeeping or cleanup, etc. associated with hwloc. Then that would simplify the code in netdetect. Maybe netdetect would still need to know about the C hwloc types, but it could call into the hwloc bindings, and then the bindings would know what needs to be mutexed and what doesn't.

Reply:
I see where you are going with that.  That seems much a larger change than I wanted to make at this time.  At the moment, we only know of this remaining concurrency problem, and it's resolved with a simple mutex around the offending call.  I could make the structural changes to build complete Go bindings for hwloc that encapsulate protection where needed, but that seems like a lot of work for something that we might not need.  I did create DAOS-5572 to look at hwloc source code to see if I might need additional protection.  If the answer to that investigation is "yes, there are more concurrency issues lurking" then I think it would be a good investment of time to create the Go bindings as you suggested.  At any rate, that's just my perspective.  If you and @mjmac want me to create those Go bindings now as part of this PR, I will do that.

Reply:
I'm also of the opinion that it would make this code easier to understand, to separate out the lower-level hwloc logic from the real business logic of netdetect. The way you pulled out the hwloc topology init logic is a really good example. Maybe it could be accomplished in a follow-on PR, if you don't want to do it now. I'll leave it up to you and @mjmac to decide, but I do think it would help in maintainability.

Reply:
I'd like feedback from @mjmac before I do any further structural changes in this PR.  I did create DAOS-5572 to investigate if additional locking would be required elsewhere.  If there is additional need, it would make sense to do any refactoring at that time.  I'm not certain that it is worthwhile to do that degree of change for this particular PR.   That said, if that is deemed necessary right now, then I'll do it.